File: 236. Lowest Common Ancestor of a Binary Tree.cpp
//Solution1:
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<TreeNode*> path1, path2;
        path1.push_back(root), path2.push_back(root);
        search(p, root, path1);
        search(q, root, path2);
        while (path1.size() > path2.size()) {
            path1.pop_back();
        }
        while (path2.size() > path1.size()) {
            path2.pop_back();
        }
        while (path1.back() != path2.back()) {
            path1.pop_back();
            path2.pop_back();
        }
        return path1.back();
    }

    bool search(TreeNode* &target, TreeNode* head, vector<TreeNode*> &path) {
        if (target == head)
            return true;
        if (!head->left && !head->right)
            return false;
        if (head->left) {
            path.push_back(head->left);
            if (search(target, head->left, path))
                return true;
            path.pop_back();
        }
        if (head->right) {
            path.push_back(head->right);
            if (search(target, head->right, path))
                return true;
            path.pop_back();
        }
        return false;
    }
};

//Solution2
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
        if (root == p || root == q || root == NULL) return root;
        TreeNode *left = lowestCommonAncestor(root->left, p, q), *right = lowestCommonAncestor(root->right, p, q);
        return left && right ? root : left ? left : right;
    }
}

File: java/236. Lowest Common Ancestor of a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// O(n)
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Deque<TreeNode> dqA = new ArrayDeque<>();
        Deque<TreeNode> dqB = new ArrayDeque<>();
        dfs(root, p, dqA);
        dfs(root, q, dqB);
        while (dqA.size() > dqB.size()) {
            dqA.pollLast();
        }
        while (dqB.size() > dqA.size()) {
            dqB.pollLast();
        }
        while (dqA.peekLast() != dqB.peekLast()) {
            dqA.pollLast();
            dqB.pollLast();
        }
        return dqA.peekLast();
    }

    private boolean dfs(TreeNode root, TreeNode target, Deque<TreeNode> dq) {
        if (root == null) {
            return false;
        }
        dq.offer(root);
        if (root == target || dfs(root.left, target, dq) || dfs(root.right, target, dq)) {
            return true;
        }
        dq.pollLast();
        return false;
    }
}


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Map<TreeNode, TreeNode> parent = new HashMap<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        parent.put(root, null);
        stack.push(root);
        while (!parent.containsKey(p) || !parent.containsKey(q)) {
            TreeNode node = stack.pop();
            if (node.left != null) {
                parent.put(node.left, node);
                stack.push(node.left);
            }
            if (node.right != null) {
                parent.put(node.right, node);
                stack.push(node.right);
            }
        }
        Set<TreeNode> ancestors = new HashSet<>();
        while (p != null) {
            ancestors.add(p);
            p = parent.get(p);
        }
        while (!ancestors.contains(q)) {
            q = parent.get(q);
        }
        return q;
    }
}
